subprojects/solver/solver-smtlib/src/main/java/hu/bme/mit/theta/solver/smtlib/impl/generic/GenericSmtLibTermTransformer.java

```diff
@@ -723,16 +722,14 @@ protected Type transformSort(final SortContext ctx) {
     /* Variable scope handling */
 
     protected void pushParams(final List<ParamDecl<? extends Type>> paramDecls,
-                              BiMap<ParamDecl<?>, String> vars) {
-        vars.putAll(paramDecls.stream()
-                .collect(Collectors.toUnmodifiableMap(Function.identity(), Decl::getName)));
+                              Stack<ParamDecl<?>> vars) {
+        vars.push();
+        vars.add(paramDecls);
     }
 
     protected void popParams(final List<ParamDecl<? extends Type>> paramDecls,
-                             BiMap<ParamDecl<?>, String> vars) {
-        for (final var paramDecl : paramDecls) {
-            vars.remove(paramDecl, paramDecl.getName());
-        }
+                             Stack<ParamDecl<?>> vars) {
+        vars.pop();
```

Comment:
Is it true, that we always pop _exactly_ what we pushed last time? If so, I can simplify this method.

Fyi, this is why we need this: in smtlib, rebinding `let`-params is apparently legal:
`(let ((x 1)) (= (let ((x 0)) x) x))`

Reply:
Up until this point we did not really support `let`, only in some cases. `pushParams` and `popParams` are used for function arguments, `for-all`, `exists`, etc... For this reason, I think we _do_ pops what we pushed last time.

Reply:
Great, thanks!